File: src/types/agents.ts
/**
 * Agent Types - TypeScript types for agent integration
 * Matches Rust backend types via Tauri invoke
 */

// ═══════════════════════════════════════════════════════════════════════════════
// CONTEXT TYPES (what agents can "see")
// ═══════════════════════════════════════════════════════════════════════════════

/** Context from the Script Editor */
export interface ScriptContext {
  /** Full script content as plain text */
  full_text: string;
  /** Currently selected text */
  selection?: string;
  /** Cursor line number (1-indexed) */
  cursor_line?: number;
  /** Current scene heading */
  current_scene?: string;
  /** Characters in current scene */
  scene_characters: string[];
  /** Current element type (action, dialogue, etc.) */
  current_element?: string;
}

/** Context from the Canvas */
export interface CanvasContext {
  /** Selected node IDs */
  selected_nodes: string[];
  /** Selected node types */
  selected_types: string[];
  /** Description of selection */
  selection_description?: string;
  /** Current zoom level */
  zoom: number;
  /** Viewport center */
  viewport_center: [number, number];
}

/** Context from the Timeline */
export interface TimelineContext {
  /** Playhead position in seconds */
  playhead: number;
  /** Selected clip IDs */
  selected_clips: string[];
  /** Active track */
  active_track?: string;
  /** Total duration */
  total_duration: number;
}

/** Token summary from vault */
export interface TokenSummary {
  id: string;
  name: string;
  description: string;
  has_reference_images: boolean;
  has_lora: boolean;
}

/** Vault context (tokens) */
export interface VaultTokenContext {
  characters: TokenSummary[];
  locations: TokenSummary[];
  props: TokenSummary[];
  style_notes?: string;
}

/** Complete context for agent request */
export interface AgentContext {
  script?: ScriptContext;
  canvas?: CanvasContext;
  timeline?: TimelineContext;
  vault?: VaultTokenContext;
  mode: 'writer' | 'studio';
  project_name?: string;
}

// ═══════════════════════════════════════════════════════════════════════════════
// ACTION TYPES (what agents can "do")
// ═══════════════════════════════════════════════════════════════════════════════

export type AudioActionType = 'Voice' | 'Music' | 'SoundEffect' | 'Ambient';
export type ScriptUpdateMode = 'Replace' | 'Insert' | 'ReplaceSelection' | 'Patch';
export type CanvasNodeType = 'Image' | 'Video' | 'Character' | 'Location' | 'Prop' | 'Note' | 'Reference';

export type AgentAction =
  | { type: 'GenerateImage'; prompt: string; model: string; width: number; height: number; token_ids: string[] }
  | { type: 'GenerateVideo'; prompt: string; model: string; duration_seconds: number; reference_image?: string; token_ids: string[] }
  | { type: 'GenerateAudio'; prompt: string; audio_type: AudioActionType; model: string; duration_seconds?: number }
  | { type: 'UpdateScript'; mode: ScriptUpdateMode; content: string; line_start?: number; line_end?: number }
  | { type: 'AddToCanvas'; node_type: CanvasNodeType; content: string; position?: [number, number]; token_id?: string }
  | { type: 'UpdateVault'; token_type: string; token_name: string; description: string; reference_image?: string }
  | { type: 'Delegate'; target_agent: string; message: string }
  | { type: 'ShowMessage'; title: string; content: string; suggestions: string[] }
  | { type: 'ExecuteWorkflow'; workflow_json: string };

// ═══════════════════════════════════════════════════════════════════════════════
// API TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export interface ChatMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface FullAgentRequest {
  agent_role: AgentRole;
  message: string;
  context?: AgentContext;
  history: ChatMessage[];
  provider?: string;
  model?: string;
  auto_execute: boolean;
}

export interface ActionResult {
  success: boolean;
  action_type: string;
  execution_id?: string;
  data?: string; // JSON string, parse as needed
  error?: string;
  credits_used?: number;
}

export interface FullAgentResponse {
  message: string;
  agent_role: string;
  model_used: string;
  actions: AgentAction[];
  action_results: ActionResult[];
  tokens_used?: number;
}

// ═══════════════════════════════════════════════════════════════════════════════
// AGENT ROLES
// ═══════════════════════════════════════════════════════════════════════════════

export type AgentRole =
  | 'showrunner'
  | 'scriptwriter'
  | 'cinematographer'
  | 'casting_director'
  | 'art_director'
  | 'voice_actors'
  | 'music_sfx_director'
  | 'photography_director'
  | 'camera_director'
  | 'editor'
  | 'colorist';

export const AGENT_ROLES: Record<AgentRole, { name: string; description: string; icon: string }> = {
  showrunner: {
    name: 'The Showrunner',
    description: 'Guardian of consistency and creative vision',
    icon: '🎬',
  },
  scriptwriter: {
    name: 'Scriptwriter',
    description: 'Screenplay, dialogue, and plot',
    icon: '✍️',
  },
  cinematographer: {
    name: 'Cinematographer',
    description: 'Lenses, lighting, camera angles',
    icon: '🎥',
  },
  casting_director: {
    name: 'Casting Director',
    description: 'Character consistency and FaceID',
    icon: '👤',
  },
  art_director: {
    name: 'Art Director',
    description: 'Locations, sets, props',
    icon: '🎨',
  },
  voice_actors: {
    name: 'Voice Actors',
    description: 'TTS and dialogue performance',
    icon: '🎙️',
  },
  music_sfx_director: {
    name: 'Music & SFX',
    description: 'Score, foley, sound design',
    icon: '🎵',
  },
  photography_director: {
    name: 'Photography Director',
    description: 'Image generation',
    icon: '📷',
  },
  camera_director: {
    name: 'Camera Director',
    description: 'Video generation',
    icon: '🎞️',
  },
  editor: {
    name: 'Editor',
    description: 'Montage, pacing, assembly',
    icon: '✂️',
  },
  colorist: {
    name: 'Colorist',
    description: 'Color grading and LUTs',
    icon: '🌈',
  },
};

File: src-tauri/src/commands/agents.rs
//! Agent Commands - Context-aware agent chat with action execution
//!
//! Replaces the basic agent_chat with full context and action support.

use serde::{Deserialize, Serialize};
use specta::Type;

use crate::ai::{
    actions::{parse_actions_from_response, ActionExecutor, ActionResult, AgentAction},
    agent_executor::{get_agent_executor, ChatMessage},
    context::AgentContext,
};

// ═══════════════════════════════════════════════════════════════════════════════
// REQUEST/RESPONSE TYPES
// ═══════════════════════════════════════════════════════════════════════════════

/// Full agent chat request with rich context
#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct FullAgentRequest {
    /// Agent role (showrunner, scriptwriter, etc.)
    pub agent_role: String,
    /// User's message
    pub message: String,
    /// Rich context from the software
    pub context: Option<AgentContext>,
    /// Conversation history
    pub history: Vec<ChatMessage>,
    /// LLM provider override
    pub provider: Option<String>,
    /// Model override
    pub model: Option<String>,
    /// Auto-execute actions?
    pub auto_execute: bool,
}

/// Full agent response with actions
#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct FullAgentResponse {
    /// Agent's text response
    pub message: String,
    /// Which agent responded
    pub agent_role: String,
    /// Model used
    pub model_used: String,
    /// Detected actions from response
    pub actions: Vec<AgentAction>,
    /// Execution results (if auto_execute was true)
    pub action_results: Vec<ActionResult>,
    /// Token usage
    pub tokens_used: Option<u32>,
}

// ═══════════════════════════════════════════════════════════════════════════════
// TAURI COMMANDS
// ═══════════════════════════════════════════════════════════════════════════════

/// Chat with an agent with full context support
#[tauri::command]
#[specta::specta]
pub async fn agent_chat_full(request: FullAgentRequest) -> Result<FullAgentResponse, String> {
    // Build context string
    let context_str = request
        .context
        .as_ref()
        .map(|c| c.to_prompt_context())
        .filter(|s| !s.is_empty());

    // Call the agent executor
    let executor = get_agent_executor();
    let chat_request = crate::ai::agent_executor::AgentChatRequest {
        agent_role: request.agent_role.clone(),
        message: request.message,
        context: context_str,
        history: request.history,
        provider: request.provider,
        model: request.model,
    };

    let response = executor.chat(chat_request).await?;

    // Parse actions from response
    let actions = parse_actions_from_response(&response.message);

    // Execute actions if requested
    let action_results = if request.auto_execute && !actions.is_empty() {
        let mut results = Vec::new();
        for action in &actions {
            let result = ActionExecutor::execute(action.clone()).await;
            results.push(result);
        }
        results
    } else {
        Vec::new()
    };

    Ok(FullAgentResponse {
        message: response.message,
        agent_role: request.agent_role,
        model_used: response.model_used,
        actions,
        action_results,
        tokens_used: response.tokens_used,
    })
}

/// Execute a single action
#[tauri::command]
#[specta::specta]
pub async fn execute_agent_action(action: AgentAction) -> Result<ActionResult, String> {
    Ok(ActionExecutor::execute(action).await)
}

/// Execute multiple actions
#[tauri::command]
#[specta::specta]
pub async fn execute_agent_actions(actions: Vec<AgentAction>) -> Result<Vec<ActionResult>, String> {
    let mut results = Vec::new();
    for action in actions {
        results.push(ActionExecutor::execute(action).await);
    }
    Ok(results)
}

/// Route a message to the best agent
#[tauri::command]
#[specta::specta]
pub fn route_message_to_agent(message: String) -> String {
    let executor = get_agent_executor();
    let role = executor.route_request(&message);
    format!("{:?}", role).to_lowercase()
}

/// Get list of agent roles
#[tauri::command]
#[specta::specta]
pub fn get_agent_roles() -> Vec<String> {
    vec![
        "showrunner".into(),
        "scriptwriter".into(),
        "cinematographer".into(),
        "casting_director".into(),
        "art_director".into(),
        "voice_actors".into(),
        "music_sfx_director".into(),
        "photography_director".into(),
        "camera_director".into(),
        "editor".into(),
        "colorist".into(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ai::context::ScriptContext;

    #[test]
    fn test_request_with_context() {
        let request = FullAgentRequest {
            agent_role: "scriptwriter".into(),
            message: "Improve this dialogue".into(),
            context: Some(AgentContext {
                script: Some(ScriptContext {
                    full_text: "JOHN: Hello".into(),
                    selection: Some("Hello".into()),
                    cursor_line: Some(1),
                    current_scene: None,
                    scene_characters: vec!["JOHN".into()],
                    current_element: Some("dialogue".into()),
                }),
                canvas: None,
                timeline: None,
                vault: None,
                mode: "writer".into(),
                project_name: Some("Test".into()),
            }),
            history: vec![],
            provider: None,
            model: None,
            auto_execute: false,
        };

        assert_eq!(request.agent_role, "scriptwriter");
        assert!(request.context.is_some());
    }
}
